File: src/device/leap_motion.cpp
#include "leap_motion.hpp"

namespace trame
{

leap_motion::leap_motion()
{

}

leap_motion::leap_motion(const leap_motion &l)
{

}

leap_motion::~leap_motion()
{

}

skeleton leap_motion::get()
{
    skeleton s = skeleton::default_skeleton();

    // get left and right wrist joints from default skeleton
    joint left_wrist, right_wrist;
    left_wrist = s.get_joint(joint_type::WRIST_LEFT);
    right_wrist = s.get_joint(joint_type::WRIST_RIGHT);
    // build left hand with leap
    joint left_hand = adapter.get_left_hand(left_wrist);
    left_wrist.valid = left_hand.valid;
    left_wrist.update(joint_type::HAND_LEFT, std::move(left_hand));
    // build right hand with leap
    joint right_hand = adapter.get_right_hand(right_wrist);
    right_wrist.valid = right_hand.valid;
    right_wrist.update(joint_type::HAND_RIGHT, std::move(right_hand));

    // update skeleton
    s.update_joint(joint_type::WRIST_LEFT, left_wrist);
    s.update_joint(joint_type::WRIST_RIGHT, right_wrist);

    s.id = 0;
    s.timestamp = current_time();
    s.valid = left_wrist.valid && right_wrist.valid;

    return s;
}

}

File: csharp/Trame/Implementation/Device/LeapMotion.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Trame.Implementation.Device.Adapter;
using Trame.Implementation.Skeleton;

namespace Trame.Implementation.Device
{
    class LeapMotion : IDevice
    {
        readonly LeapAdapter adapter = new LeapAdapter();
        private Thread t;
        private bool running = true;

        public LeapMotion()
        {
            t = new Thread(Run);
            t.Start();
        }
        
        private void Run()
        {
            while (running)
            {
                FireNewSkeleton();
            }
        }

        public ISkeleton GetSkeleton()
        {
            ISkeleton s = Creator.GetNewDefaultSkeleton();
            return GetSkeleton(s);
        }

        public ISkeleton GetSkeleton(ISkeleton s)
        {
            IJoint leftWrist = s.GetJoint(JointType.WRIST_LEFT);
            IJoint rightWrist = s.GetJoint(JointType.WRIST_RIGHT);

            IJoint leftHand = adapter.LeftHand(leftWrist);
            leftWrist.Valid = leftHand.Valid;
            leftWrist.Update(JointType.HAND_LEFT, leftHand);

            IJoint rightHand = adapter.RightHand(rightWrist);
            rightWrist.Valid = rightHand.Valid;
            rightWrist.Update(JointType.HAND_RIGHT, rightHand);

            s.UpdateSkeleton(JointType.WRIST_LEFT, leftWrist);
            s.UpdateSkeleton(JointType.WRIST_RIGHT, rightWrist);

            s.Valid = leftWrist.Valid && rightHand.Valid;

            return s;
        }

        public void Stop()
        {
            adapter.Stop();
            running = false;
            t.Join();
        }

        private void FireNewSkeleton()
        {
            if (NewSkeleton != null)
            {
                NewSkeleton(GetSkeleton());
            }
        }

        public event Action<ISkeleton> NewSkeleton;
    }
}
